client/consensus/grandpa/environment.go

```diff
@@ -0,0 +1,570 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"fmt"
+	"github.com/ChainSafe/gossamer/lib/crypto/ed25519"
+	grandpa "github.com/ChainSafe/gossamer/pkg/finality-grandpa"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"golang.org/x/exp/constraints"
+	"golang.org/x/exp/slices"
+	"sync"
+)
+
+// completedRound Data about a completed round. The set of votes that is stored must be
+// minimal, i.e. at most one equivocation is stored per voter.
+type completedRound[H comparable, N constraints.Unsigned] struct {
+	// The round number
+	Number uint64
+	// The round state (prevote ghost, estimate, finalized, etc.)
+	State grandpa.RoundState[H, N]
+	// The target block base used for voting in the round
+	Base grandpa.HashNumber[H, N]
+	// All the votes observed in the round
+	// I think this is signature type, double check
+	Votes []grandpa.SignedMessage[H, N, ed25519.SignatureBytes, ed25519.PublicKey]
+}
+
+// NumLastCompletedRounds NOTE: the current strategy for persisting completed rounds is very naive
+// (update everything) and we also rely on cloning to do atomic updates,
+// therefore this value should be kept small for now.
+const NumLastCompletedRounds = 2
+
+// completedRounds Data about last completed rounds within a single voter set. Stores
+// NumLastCompletedRounds and always contains data about at least one round
+// (genesis).
+type completedRounds[H comparable, N constraints.Unsigned] struct {
+	Rounds []completedRound[H, N]
+	SetId  uint64
+	Voters []ed25519.PublicKey
+}
+
+// NewCompletedRounds Create a new completed rounds tracker with NUM_LAST_COMPLETED_ROUNDS capacity.
+func NewCompletedRounds[H comparable, N constraints.Unsigned](genesis *completedRound[H, N], setId uint64, voters AuthoritySet[H, N]) completedRounds[H, N] {
+	rounds := make([]completedRound[H, N], 0, NumLastCompletedRounds)
+	if genesis != nil {
+		rounds = append(rounds, *genesis)
+	}
+
+	var voterIDs []ed25519.PublicKey
+	currentAuthorities := voters.CurrentAuthorities
+	for _, auth := range currentAuthorities {
+		voterIDs = append(voterIDs, auth.Key)
+	}
+
+	return completedRounds[H, N]{
+		rounds,
+		setId,
+		voterIDs,
+	}
+}
+
+// last Returns the last (latest) completed round
+func (compRounds *completedRounds[H, N]) last() completedRound[H, N] {
+	if len(compRounds.Rounds) == 0 {
+		panic("inner is never empty; always contains at least genesis; qed")
+	}
+	return compRounds.Rounds[0]
+}
+
+// push a new completed round, oldest round is evicted if number of rounds
+// is higher than `NUM_LAST_COMPLETED_ROUNDS`.
+func (compRounds *completedRounds[H, N]) push(compRound completedRound[H, N]) {
```

Comment:
go best practice to use abbreviations
```suggestion
func (cr *completedRounds[H, N]) push(compRound completedRound[H, N]) {
```